### PR TITLE
[DLFL21] fix errors on running 15-transformers.ipynb

### PR DESCRIPTION
**Pull request by NoTody**
[DLFL21] 1. E.requires_grad=False need to be set before the operations; otheriwise, "runtimeerror: a view of a leaf variable that requires grad is being used in an in-place operation." will occur when initializing the model. 2. torchtext.legacy need to used when retrieving IMDB dataset in newest version of torch.

### CODE REVIEW

**Review by Atcold — CHANGES_REQUESTED** [review]
Can you remove all these
```
"metadata": {
    "id": "U_mTDKQcmOxc"
   },
```

File: 15-transformer.ipynb [hunk 1]
```diff
@@ -2,16 +2,20 @@
  "cells": [
   {
    "cell_type": "markdown",
-   "metadata": {},
+   "metadata": {
+    "id": "U_mTDKQcmOxc"
+   },
```
Comment (Atcold) [inline 15-transformer.ipynb:7]: Remove.

File: 15-transformer.ipynb [hunk 2]
```diff
@@ -2,16 +2,20 @@
...
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
-   "metadata": {},
+   "execution_count": 34,
+   "metadata": {
+    "id": "nXTum9GcmOxg"
+   },
```
Comment (Atcold) [inline 15-transformer.ipynb:18]: Revert.

File: 15-transformer.ipynb [hunk 3]
```diff
@@ -22,8 +26,10 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
-   "metadata": {},
+   "execution_count": 35,
+   "metadata": {
+    "id": "QyvquvmlmOxh"
+   },
```
Comment (Atcold) [inline 15-transformer.ipynb:32]: Revert.

File: 15-transformer.ipynb [hunk 4]
```diff
@@ -138,15 +152,22 @@
...
   {
    "cell_type": "code",
-   "execution_count": 5,
+   "execution_count": 38,
    "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "bWGgOyqWmOxk",
+    "outputId": "b760f8ce-d258-4c23-f240-a255e6510d3a",
```
Comment (Atcold) [inline 15-transformer.ipynb:170]: Revert

File: 15-transformer.ipynb [hunk 5]
```diff
@@ -555,43 +620,13 @@
...
-       "          (W_h): Linear(in_features=32, out_features=32, bias=True)\n",
-       "        )\n",
-       "        (cnn): CNN(\n",
-       "          (k1convL1): Linear(in_features=32, out_features=128, bias=True)\n",
-       "          (k1convL2): Linear(in_features=128, out_features=32, bias=True)\n",
-       "          (activation): ReLU()\n",
-       "        )\n",
-       "        (layernorm1): LayerNorm((32,), eps=1e-06, elementwise_affine=True)\n",
-       "        (layernorm2): LayerNorm((32,), eps=1e-06, elementwise_affine=True)\n",
-       "      )\n",
```
Comment (Atcold) [inline 15-transformer.ipynb:587]: What happened?

File: 15-transformer.ipynb (hunk 1 above)
Comment (NoTody) [reply 15-transformer.ipynb]: In the new commits, I only change the lines need to be changed from original file without executing any blocks. I think this one would only include the commits you need.

**Review by NoTody — COMMENTED** [review]
In the new commit, I only change the lines need to be changed from original file without executing any blocks. I think these only changed 12 lines would be the needed commits.

File: 15-transformer.ipynb [hunk 6]
```diff
@@ -750,7 +748,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 [conda env:pDL]",
+   "display_name": "Python 3",
```
Comment (Atcold) [inline 15-transformer.ipynb:751]:
```suggestion
   "display_name": "Python 3 [conda env:pDL]",
```

File: 15-transformer.ipynb [hunk 7]
```diff
@@ -764,7 +762,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.2"
+   "version": "3.7.3"
```
Comment (Atcold) [inline 15-transformer.ipynb:765]:
```suggestion
   "version": "3.8.2"
```

File: 15-transformer.ipynb [hunk 8]
```diff
@@ -448,13 +449,10 @@
    ],
    "source": [
     "max_len = 200\n",
-    "text = data.Field(sequential=True, fix_length=max_len, batch_first=True, lower=True, dtype=torch.long)\n",
-    "label = data.LabelField(sequential=False, dtype=torch.long)\n",
-    "datasets.IMDB.download('./')\n",
-    "ds_train, ds_test = datasets.IMDB.splits(text, label, path='./imdb/aclImdb/')\n",
-    "print('train : ', len(ds_train))\n",
-    "print('test : ', len(ds_test))\n",
-    "print('train.fields :', ds_train.fields)"
```
Comment (Atcold) [inline 15-transformer.ipynb:457]: Why did you remove these?

File: 15-transformer.ipynb [hunk 9]
```diff
@@ -354,10 +354,10 @@
     "        [p / np.power(10000, 2 * (j // 2) / dim) for j in range(dim)]\n",
     "        for p in range(nb_p)\n",
     "    ])\n",
-    "    E[:, 0::2] = torch.FloatTensor(np.sin(theta[:, 0::2]))\n",
-    "    E[:, 1::2] = torch.FloatTensor(np.cos(theta[:, 1::2]))\n",
     "    E.detach_()\n",
     "    E.requires_grad = False\n",
+    "    E[:, 0::2] = torch.FloatTensor(np.sin(theta[:, 0::2]))\n",
+    "    E[:, 1::2] = torch.FloatTensor(np.cos(theta[:, 1::2]))\n",
```
Comment (Atcold) [inline 15-transformer.ipynb:360]: Why have these lines been moved around?

**Review by Atcold — CHANGES_REQUESTED** [review]

File: 15-transformer.ipynb (hunk 9 above)
Comment (NoTody) [reply 15-transformer.ipynb:360]: error
Not sure about your case, this would give me this error in the model creation block.

File: 15-transformer.ipynb (hunk 8 above)
Comment (NoTody) [reply 15-transformer.ipynb:457]: So this way of importing data no long works in newest pytorch. If you try to do it in this way, it would give this error.
image
Instead, torchtext.legacy should be used if you want to achieve the same functionality.

Comment (NoTody) [reply 15-transformer.ipynb:457]: Oh, I just moved it to next block because of my personal convention of writing in jupyter. I will move it back.

Comment (NoTody) [reply 15-transformer.ipynb:457]: I will add these back. I just forgot to add them back when changing this block..

File: 15-transformer.ipynb [hunk 10]
```diff
@@ -448,10 +449,9 @@
    ],
    "source": [
     "max_len = 200\n",
-    "text = data.Field(sequential=True, fix_length=max_len, batch_first=True, lower=True, dtype=torch.long)\n",
-    "label = data.LabelField(sequential=False, dtype=torch.long)\n",
-    "datasets.IMDB.download('./')\n",
```
Comment (Atcold) [inline 15-transformer.ipynb:453]: So, is downloading no longer necessary?

File: 15-transformer.ipynb [hunk 11]
```diff
@@ -750,7 +750,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 [conda env:pDL]",
+   "display_name": "Python 3",
```
Comment (Atcold) [inline 15-transformer.ipynb:753]:
```suggestion
   "display_name": "Python 3 [conda env:pDL]",
```

File: 15-transformer.ipynb (hunk 7 above)
Comment (Atcold) [reply 15-transformer.ipynb:767]: What's going on?

**Review by Atcold — CHANGES_REQUESTED** [review]
Still, comments to be addressed.

File: 15-transformer.ipynb [hunk 12]
```diff
@@ -354,10 +354,10 @@
     "        [p / np.power(10000, 2 * (j // 2) / dim) for j in range(dim)]\n",
     "        for p in range(nb_p)\n",
     "    ])\n",
-    "    E[:, 0::2] = torch.FloatTensor(np.sin(theta[:, 0::2]))\n",
-    "    E[:, 1::2] = torch.FloatTensor(np.cos(theta[:, 1::2]))\n",
     "    E.detach_()\n",
```
Comment (Atcold) [inline 15-transformer.ipynb:357]: This line is unnecessary, right?
In particular, we are not learning the embeddings.

File: 15-transformer.ipynb (hunk 9 above)
Comment (Atcold) [reply 15-transformer.ipynb:360]: When generating `position_embeddings = nn.Embedding(max_position_embeddings, d_model)` we should say that we do not require gradients because we don't want to learn the embeddings.

File: 15-transformer.ipynb (hunk 10 above)
Comment (NoTody) [reply 15-transformer.ipynb:453]: Yes, no need to call download here.

**Review by Atcold — APPROVED** [review]
Thanks!
